File: 5lab/cpp/Order.cpp
#include "Order.h"
#include <iostream>
#include <iomanip>
#include <cstdlib>
#include <ctime>
#include <fstream>
#include <sstream>
#include <limits>
#include <algorithm>

using namespace std;

#ifdef _MSC_VER
#define _CRT_SECURE_NO_WARNINGS
#endif

// Конструктор по умолчанию
Order::Order() : itemCount(0), countOrder(0), costOrder(0.0) {
    // Инициализация массива указателей
    for (int i = 0; i < MAX_ORDER_ITEMS; i++) {
        orderItems[i] = nullptr;
        quantities[i] = 0;
    }

    numberOrder = generateOrderNumber();
    statusOrder = "Ожидает оплаты";
    dateOrder = getCurrentDate();
    shopOrder = "Музыкальный магазин 'MusicLoverzz'";
    payWay = "Не выбран";

    cout << "Вызван конструктор по умолчанию Order: " << numberOrder << endl;
}

// Конструктор с параметрами
Order::Order(const std::string& clientId, Basket& basket) :
    clientOrder(clientId), itemCount(0), countOrder(0), costOrder(0.0) {

    // Инициализация массива указателей
    for (int i = 0; i < MAX_ORDER_ITEMS; i++) {
        orderItems[i] = nullptr;
        quantities[i] = 0;
    }

    numberOrder = generateOrderNumber();
    statusOrder = "Ожидает оплаты";
    dateOrder = getCurrentDate();
    shopOrder = "Музыкальный магазин 'MusicLoverzz'";
    payWay = "Не выбран";

    CopyProductsFromBasket(basket);

    cout << "Вызван конструктор с параметрами Order: " << numberOrder << endl;
}

// КОНСТРУКТОР КОПИРОВАНИЯ
Order::Order(const Order& other) {
    cout << "Вызван конструктор копирования Order: копируем " << other.numberOrder << endl;

    // Копируем простые поля
    numberOrder = generateOrderNumber();  // Новый номер для копии
    clientOrder = other.clientOrder;
    itemCount = other.itemCount;
    countOrder = other.countOrder;
    costOrder = other.costOrder;
    statusOrder = other.statusOrder;
    payWay = other.payWay;
    shopOrder = other.shopOrder;
    dateOrder = getCurrentDate();  // Новая дата для копии

    // Инициализация массива указателей
    for (int i = 0; i < MAX_ORDER_ITEMS; i++) {
        orderItems[i] = nullptr;
        quantities[i] = 0;
    }

    // Глубокое копирование товаров
    for (int i = 0; i < itemCount; i++) {
        if (other.orderItems[i] != nullptr) {
            // Создаем ПОЛНУЮ КОПИЮ продукта
            orderItems[i] = new Product(*(other.orderItems[i]));
            quantities[i] = other.quantities[i];
        }
    }
}

// ОПЕРАТОР ПРИСВАИВАНИЯ
Order& Order::operator=(const Order& other) {
    cout << "Вызван оператор присваивания Order: присваиваем " << other.numberOrder << endl;

    if (this != &other) {  // Защита от самоприсваивания
        // Очищаем текущие данные
        clearOrderItems();

        // Копируем простые поля
        // numberOrder НЕ копируем - у каждого заказа свой уникальный номер
        clientOrder = other.clientOrder;
        itemCount = other.itemCount;
        countOrder = other.countOrder;
        costOrder = other.costOrder;
        statusOrder = other.statusOrder;
        payWay = other.payWay;
        shopOrder = other.shopOrder;
        dateOrder = getCurrentDate();  // Обновляем дату

        // Глубокое копирование товаров
        for (int i = 0; i < itemCount; i++) {
            if (other.orderItems[i] != nullptr) {
                orderItems[i] = new Product(*(other.orderItems[i]));
                quantities[i] = other.quantities[i];
            }
        }
    }

    return *this;
}

// Деструктор
Order::~Order() {
    cout << "Вызван деструктор Order: " << numberOrder << endl;
    clearOrderItems();
}

// Очистка массива товаров
void Order::clearOrderItems() {
    for (int i = 0; i < itemCount; i++) {
        if (orderItems[i] != nullptr) {
            delete orderItems[i];
            orderItems[i] = nullptr;
        }
    }
    itemCount = 0;
}

// Копирование товаров из корзины
void Order::CopyProductsFromBasket(Basket& basket) {
    if (basket.IsEmpty()) {
        return;
    }

    itemCount = basket.getProductCount();
    countOrder = basket.GetBasketCount();
    costOrder = basket.GetBasketCost();

    for (int i = 0; i < itemCount; i++) {
        Product* basketProduct = basket.getProduct(i);
        if (basketProduct != nullptr) {
            // Создаем ПОЛНУЮ КОПИЮ товара
            orderItems[i] = new Product(
                basketProduct->getArticle(),
                basketProduct->getName(),
                basketProduct->getBrand(),
                basketProduct->getProducerCountry(),
                basketProduct->getCategory(),
                basketProduct->getTypeProduct(),
                basketProduct->getDescription(),
                basketProduct->getQuantityStock(),
                basketProduct->getPrice()
            );
            quantities[i] = basket.getQuantity(i);
        }
    }
}

// Глубокое копирование
void Order::deepCopy(const Order& other) {
    clearOrderItems();

    // Копируем простые поля
    clientOrder = other.clientOrder;
    itemCount = other.itemCount;
    countOrder = other.countOrder;
    costOrder = other.costOrder;
    statusOrder = other.statusOrder;
    payWay = other.payWay;
    shopOrder = other.shopOrder;

    // Глубокое копирование товаров
    for (int i = 0; i < itemCount; i++) {
        if (other.orderItems[i] != nullptr) {
            orderItems[i] = new Product(*(other.orderItems[i]));
            quantities[i] = other.quantities[i];
        }
    }
}

// Статический метод для создания заказа
bool Order::MakeOrder(const std::string& clientId, Basket& basket) {
    if (basket.IsEmpty()) {
        cout << "Корзина пуста! Добавьте товары перед оформлением заказа." << endl;
        return false;
    }

    char confirmation;
    cout << "\n=== ПОДТВЕРЖДЕНИЕ ЗАКАЗА ===" << endl;
    basket.DisplayBasket();
    cout << "\nВы точно хотите оформить заказ? (y/n): ";
    cin >> confirmation;

    if (confirmation != 'y' && confirmation != 'Y') {
        cout << "Оформление заказа отменено." << endl;
        return false;
    }

    Order* newOrder = new Order(clientId, basket);

    if (newOrder->Payment()) {
        newOrder->ChangeStatus("Оплачен");
        cout << "\n=== ЗАКАЗ УСПЕШНО ОФОРМЛЕН ===" << endl;
        newOrder->InfoOrder();
        newOrder->SaveToFile();
        basket.ClearBasket();
        delete newOrder;
        return true;
    }
    else {
        newOrder->ChangeStatus("Ошибка оплаты");
        cout << "Ошибка при оформлении заказа!" << endl;
        delete newOrder;
        return false;
    }
}

// Сохранение заказа в файл
void Order::SaveToFile() {
    ofstream file("orders.txt", ios::app);
    if (file.is_open()) {
        file << numberOrder << "|" << clientOrder << "|" << statusOrder << "|" << costOrder;
        for (int i = 0; i < itemCount; i++) {
            if (orderItems[i] != nullptr) {
                file << "|" << orderItems[i]->getArticle() << "|" << quantities[i];
            }
        }
        file << endl;
        file.close();
    }
}

// Изменение статуса заказа
void Order::ChangeStatus(const std::string& newStatus) {
    statusOrder = newStatus;
    if (newStatus == "Оплачен") {
        cout << "\n=== ВАШ ЗАКАЗ УСПЕШНО ОФОРМЛЕН! ===" << endl;
        cout << "Номер заказа: " << numberOrder << endl;
        cout << "Статус: " << statusOrder << endl;
        cout << "==================================" << endl;
    }
    else {
        cout << "Статус заказа изменен на: " << statusOrder << endl;
    }
}

// Оплата заказа
bool Order::Payment() {
    cout << "\n=== ОПЛАТА ЗАКАЗА ===" << endl;
    cout << "Сумма к оплате: " << costOrder << " руб." << endl;

    int paymentChoice;
    bool validChoice = false;

    while (!validChoice) {
        cout << "Выберите способ оплаты:" << endl;
        cout << "1. Банковская карта" << endl;
        cout << "2. Электронный кошелек" << endl;
        cout << "3. Наложенный платеж" << endl;
        cout << "Ваш выбор: ";
        cin >> paymentChoice;

        if (cin.fail()) {
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            cout << "Ошибка! Введите число от 1 до 3." << endl;
            continue;
        }

        if (paymentChoice >= 1 && paymentChoice <= 3) {
            validChoice = true;
        }
        else {
            cout << "Неверный выбор! Введите 1, 2 или 3." << endl;
        }
    }

    switch (paymentChoice) {
    case 1:
        payWay = "Банковская карта";
        break;
    case 2:
        payWay = "Электронный кошелек";
        break;
    case 3:
        payWay = "Наложенный платеж";
        break;
    }

    cout << "Обработка платежа..." << endl;
    cout << "Способ оплаты: " << payWay << endl;
    cout << "Оплата прошла успешно!" << endl;
    return true;
}

// Вывод информации о заказе
void Order::InfoOrder() const {
    cout << "\n=== ИНФОРМАЦИЯ О ЗАКАЗЕ ===" << endl;
    cout << "Номер заказа: " << numberOrder << endl;
    cout << "Клиент: " << clientOrder << endl;
    cout << "Дата заказа: " << dateOrder << endl;
    cout << "Статус: " << statusOrder << endl;
    cout << "Способ оплаты: " << payWay << endl;
    cout << "Количество товаров: " << countOrder << endl;
    cout << "Общая стоимость: " << costOrder << " руб." << endl;
    cout << "Магазин: " << shopOrder << endl;

    if (itemCount > 0) {
        cout << "\nСостав заказа:" << endl;
        cout << "------------------------------------" << endl;
        for (int i = 0; i < itemCount; i++) {
            if (orderItems[i] != nullptr) {
                cout << (i + 1) << ". " << orderItems[i]->getName()
                    << " - " << quantities[i] << " шт. x "
                    << orderItems[i]->getPrice() << " руб. = "
                    << (quantities[i] * orderItems[i]->getPrice()) << " руб." << endl;
            }
        }
        cout << "------------------------------------" << endl;
    }
    cout << "==================================" << endl;
}

// Генерация номера заказа
std::string Order::generateOrderNumber() {
    srand(static_cast<unsigned int>(time(0)));
    int randomNum = rand() % 10000;
    return "ORD" + std::to_string(randomNum);
}

// Получение текущей даты
std::string Order::getCurrentDate() {
    time_t now = time(0);

#ifdef _MSC_VER
    struct tm timeinfo;
    localtime_s(&timeinfo, &now);
    char buffer[80];
    strftime(buffer, 80, "%d.%m.%Y %H:%M:%S", &timeinfo);
#else
    tm* localTime = localtime(&now);
    char buffer[80];
    strftime(buffer, 80, "%d.%m.%Y %H:%M:%S", localTime);
#endif

    return std::string(buffer);
}

File: 5lab/cs/Order.cs
using System;
using System.IO;
using System.Text;
using System.Collections.Generic;

namespace MusicShopSystem
{
    public class Order : IDisposable
    {
        // Константы
        private const int MAX_ORDER_ITEMS = 100; // максимальное количество различных товаров в заказе

        // Поля класса
        private string numberOrder;              // номер заказа
        private string clientOrder;              // идентификатор клиента
        private Product[] orderItems;           // массив товаров в заказе
        private int[] quantities;               // массив количеств для каждого товара
        private int itemCount;                  // количество различных товаров в заказе
        private int countOrder;                 // общее количество
        private double costOrder;               // общая стоимость
        private string statusOrder;             // статус заказа
        private string payWay;                  // способ оплаты
        private string shopOrder;               // магазин для получения
        private string dateOrder;               // дата и время создания заказа

        private bool disposed = false;          // Флаг для отслеживания освобождения ресурсов

        // Конструктор по умолчанию
        public Order()
        {
            itemCount = 0;
            countOrder = 0;
            costOrder = 0.0;
            orderItems = new Product[MAX_ORDER_ITEMS];
            quantities = new int[MAX_ORDER_ITEMS];

            for (int i = 0; i < MAX_ORDER_ITEMS; i++)
            {
                orderItems[i] = null;
                quantities[i] = 0;
            }

            numberOrder = GenerateOrderNumber();
            statusOrder = "Ожидает оплаты";
            dateOrder = GetCurrentDate();
            shopOrder = "Музыкальный магазин 'MusicLoverzz'";
            payWay = "Не выбран";
        }

        // Конструктор с параметрами (клиент и корзина)
        public Order(string clientId, Basket basket)
        {
            clientOrder = clientId;
            itemCount = 0;
            countOrder = 0;
            costOrder = 0.0;
            orderItems = new Product[MAX_ORDER_ITEMS];
            quantities = new int[MAX_ORDER_ITEMS];

            for (int i = 0; i < MAX_ORDER_ITEMS; i++)
            {
                orderItems[i] = null;
                quantities[i] = 0;
            }

            numberOrder = GenerateOrderNumber();
            statusOrder = "Ожидает оплаты";
            dateOrder = GetCurrentDate();
            shopOrder = "Музыкальный магазин 'MusicLoverzz'";
            payWay = "Не выбран";

            CopyProductsFromBasket(basket); // копирование информации из корзины
        }


        ~Order()
        {
            Dispose(false);
        }

        // Реализация интерфейса IDisposable для явного освобождения ресурсов
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this); // Отменяем вызов финализатора
        }

        // Защищенный метод для освобождения ресурсов
        protected virtual void Dispose(bool disposing)
        {
            if (!disposed)
            {
                if (disposing)
                {
                    // Освобождаем управляемые ресурсы - товары в заказе
                    for (int i = 0; i < itemCount; i++)
                    {
                        if (orderItems[i] != null)
                        {
                            // Если Product реализует IDisposable
                            if (orderItems[i] is IDisposable disposableProduct)
                            {
                                disposableProduct.Dispose();
                            }
                            orderItems[i] = null;
                        }
                    }

                    // Обнуляем массивы
                    if (orderItems != null)
                    {
                        Array.Clear(orderItems, 0, orderItems.Length);
                    }

                    if (quantities != null)
                    {
                        Array.Clear(quantities, 0, quantities.Length);
                    }
                }

                // Освобождаем неуправляемые ресурсы (если есть)
                // В данном классе их нет

                disposed = true;
            }
        }

        // Метод копирования товаров из корзины в заказ
        private void CopyProductsFromBasket(Basket basket)
        {
            if (basket.IsEmpty()) // проверка не пустая ли корзина
            {
                return; // если корзина пуста - выходим
            }

            // Копируем основную информацию из корзины
            itemCount = basket.GetProductCount();    // количество различных товаров
            countOrder = (int)basket.GetBasketCount(); // общее количество единиц товара
            costOrder = basket.GetBasketCost();      // общая стоимость

            // Копируем каждый товар из корзины в заказ
            for (int i = 0; i < itemCount; i++)
            {
                // Получаем товар из корзины
                Product basketProduct = basket.GetProduct(i);

                // Проверяем что товар существует
                if (basketProduct != null)
                {
                    // Создаем ПОЛНУЮ КОПИЮ товара
                    orderItems[i] = new Product(
                        basketProduct.GetArticle(),
                        basketProduct.GetName(),
                        basketProduct.GetBrand(),
                        basketProduct.GetProducerCountry(),
                        basketProduct.GetCategory(),
                        basketProduct.GetTypeProduct(),
                        basketProduct.GetDescription(),
                        basketProduct.GetQuantityStock(),
                        basketProduct.GetPrice()
                    );

                    // Копируем количество товара
                    quantities[i] = basket.GetQuantity(i);
                }
            }
        }

        // Статический метод для создания заказа 
        public static bool MakeOrder(string clientId, Basket basket)
        {
            // Проверка что корзина не пустая
            if (basket.IsEmpty())
            {
                Console.WriteLine("Корзина пуста! Добавьте товары перед оформлением заказа.");
                return false; // возвращаем false если корзина пуста
            }

            // Запрос подтверждения у пользователя
            Console.WriteLine("\n=== ПОДТВЕРЖДЕНИЕ ЗАКАЗА ===");
            basket.DisplayBasket();
            Console.Write("\nВы точно хотите оформить заказ? (y/n): ");
            char confirmation = Console.ReadKey().KeyChar;
            Console.WriteLine();

            if (confirmation != 'y' && confirmation != 'Y')
            {
                Console.WriteLine("Оформление заказа отменено.");
                return false;
            }

            // Создаем новый заказ из корзины
            Order newOrder = new Order(clientId, basket);

            // Проверяем оплату
            if (newOrder.Payment())
            {
                // Если оплата прошла успешно
                newOrder.ChangeStatus("Оплачен");
                Console.WriteLine("\n=== ЗАКАЗ УСПЕШНО ОФОРМЛЕН ===");
                newOrder.InfoOrder(); // показываем информацию о заказе
                newOrder.SaveToFile(); // сохраняем заказ в файл

                // Очищаем корзину
                basket.ClearBasket();

                newOrder.Dispose(); // освобождаем ресурсы
                return true;
            }
            else
            {
                newOrder.ChangeStatus("Ошибка оплаты");
                Console.WriteLine("Ошибка при оформлении заказа!");
                newOrder.Dispose(); // освобождаем ресурсы
                return false;
            }
        }

        // Метод сохранения заказа в файл
        public void SaveToFile()
        {
            if (disposed)
                throw new ObjectDisposedException("Order", "Объект уже освобожден");

            // Открываем файл для добавления
            using (StreamWriter file = new StreamWriter("orders.txt", true, Encoding.UTF8))
            {
                // Записываем основную информацию о заказе
                file.Write($"{numberOrder}|{clientOrder}|{statusOrder}|{costOrder:F2}");

                // Записываем информацию о каждом товаре в заказе
                for (int i = 0; i < itemCount; i++)
                {
                    if (orderItems[i] != null)
                    {
                        file.Write($"|{orderItems[i].GetArticle()}|{quantities[i]}");
                    }
                }

                file.WriteLine();
            }
        }

        // Метод изменения статуса заказа
        public void ChangeStatus(string newStatus)
        {
            if (disposed)
                throw new ObjectDisposedException("Order", "Объект уже освобожден");

            statusOrder = newStatus;

            if (newStatus == "Оплачен")
            {
                Console.WriteLine("\n=== ВАШ ЗАКАЗ УСПЕШНО ОФОРМЛЕН! ===");
                Console.WriteLine($"Номер заказа: {numberOrder}");
                Console.WriteLine($"Статус: {statusOrder}");
                Console.WriteLine("==================================");
            }
            else
            {
                Console.WriteLine($"Статус заказа изменен на: {statusOrder}");
            }
        }

        // Метод обработки оплаты заказа
        public bool Payment()
        {
            if (disposed)
                throw new ObjectDisposedException("Order", "Объект уже освобожден");

            Console.WriteLine("\n=== ОПЛАТА ЗАКАЗА ===");
            Console.WriteLine($"Сумма к оплате: {costOrder:F2} руб.");

            int paymentChoice = 0; // Инициализируем значением по умолчанию
            bool validChoice = false;

            while (!validChoice)
            {
                Console.WriteLine("Выберите способ оплаты:");
                Console.WriteLine("1. Банковская карта");
                Console.WriteLine("2. Электронный кошелек");
                Console.WriteLine("3. Наложенный платеж");
                Console.Write("Ваш выбор: ");

                string input = Console.ReadLine();
                if (int.TryParse(input, out int choice))
                {
                    if (choice >= 1 && choice <= 3)
                    {
                        paymentChoice = choice; // Присваиваем значение
                        validChoice = true; // выбор корректен
                    }
                    else
                    {
                        Console.WriteLine("Неверный выбор! Введите 1, 2 или 3.");
                    }
                }
                else
                {
                    Console.WriteLine("Ошибка! Введите число от 1 до 3.");
                }
            }


            switch (paymentChoice)
            {
                case 1:
                    payWay = "Банковская карта";
                    break;
                case 2:
                    payWay = "Электронный кошелек";
                    break;
                case 3:
                    payWay = "Наложенный платеж";
                    break;
            }

            Console.WriteLine("Обработка платежа...");
            Console.WriteLine($"Способ оплаты: {payWay}");
            Console.WriteLine("Оплата прошла успешно!");

            return true;
        }

        // Метод вывода полной информации о заказе
        public void InfoOrder()
        {
            if (disposed)
                throw new ObjectDisposedException("Order", "Объект уже освобожден");

            Console.WriteLine("\n=== ИНФОРМАЦИЯ О ЗАКАЗЕ ===");
            Console.WriteLine($"Номер заказа: {numberOrder}");
            Console.WriteLine($"Клиент: {clientOrder}");
            Console.WriteLine($"Дата заказа: {dateOrder}");
            Console.WriteLine($"Статус: {statusOrder}");
            Console.WriteLine($"Способ оплаты: {payWay}");
            Console.WriteLine($"Количество товаров: {countOrder}");
            Console.WriteLine($"Общая стоимость: {costOrder:F2} руб.");
            Console.WriteLine($"Магазин: {shopOrder}");

            // Вывод состава заказа если есть товары
            if (itemCount > 0)
            {
                Console.WriteLine("\nСостав заказа:");
                Console.WriteLine("-----------------------------------");

                // Перебор всех товаров в заказе
                for (int i = 0; i < itemCount; i++)
                {
                    if (orderItems[i] != null)
                    {
                        // Вывод информации о каждом товаре
                        double itemTotal = quantities[i] * orderItems[i].GetPrice();
                        Console.WriteLine($"{i + 1}. {orderItems[i].GetName()} - " +
                                         $"{quantities[i]} шт. x {orderItems[i].GetPrice()} руб. = " +
                                         $"{itemTotal:F2} руб.");
                    }
                }

                Console.WriteLine("-----------------------------------");
            }

            Console.WriteLine("==================================");
        }

        // Метод генерации уникального номера заказа
        private string GenerateOrderNumber()
        {
            Random random = new Random();
            int randomNum = random.Next(10000); // генерация случайного числа от 0 до 9999
            return "ORD" + randomNum.ToString("D4");
        }

        // Метод получения текущей даты и времени
        private string GetCurrentDate()
        {
            return DateTime.Now.ToString("dd.MM.yyyy HH:mm:ss");
        }

        // Свойства для доступа к полям
        public string NumberOrder
        {
            get
            {
                if (disposed)
                    throw new ObjectDisposedException("Order", "Объект уже освобожден");
                return numberOrder;
            }
        }

        public string ClientOrder
        {
            get
            {
                if (disposed)
                    throw new ObjectDisposedException("Order", "Объект уже освобожден");
                return clientOrder;
            }
        }

        public string StatusOrder
        {
            get
            {
                if (disposed)
                    throw new ObjectDisposedException("Order", "Объект уже освобожден");
                return statusOrder;
            }
        }

        public double CostOrder
        {
            get
            {
                if (disposed)
                    throw new ObjectDisposedException("Order", "Объект уже освобожден");
                return costOrder;
            }
        }

        public int CountOrder
        {
            get
            {
                if (disposed)
                    throw new ObjectDisposedException("Order", "Объект уже освобожден");
                return countOrder;
            }
        }

        public string DateOrder
        {
            get
            {
                if (disposed)
                    throw new ObjectDisposedException("Order", "Объект уже освобожден");
                return dateOrder;
            }
        }

        public string PayWay
        {
            get
            {
                if (disposed)
                    throw new ObjectDisposedException("Order", "Объект уже освобожден");
                return payWay;
            }
        }

        public string ShopOrder
        {
            get
            {
                if (disposed)
                    throw new ObjectDisposedException("Order", "Объект уже освобожден");
                return shopOrder;
            }
        }

        // Методы для совместимости с C++ кодом
        public string GetNumberOrder()
        {
            if (disposed)
                throw new ObjectDisposedException("Order", "Объект уже освобожден");
            return numberOrder;
        }

        public string GetClientOrder()
        {
            if (disposed)
                throw new ObjectDisposedException("Order", "Объект уже освобожден");
            return clientOrder;
        }

        public string GetStatusOrder()
        {
            if (disposed)
                throw new ObjectDisposedException("Order", "Объект уже освобожден");
            return statusOrder;
        }

        public double GetCostOrder()
        {
            if (disposed)
                throw new ObjectDisposedException("Order", "Объект уже освобожден");
            return costOrder;
        }

        public int GetCountOrder()
        {
            if (disposed)
                throw new ObjectDisposedException("Order", "Объект уже освобожден");
            return countOrder;
        }

        public string GetDateOrder()
        {
            if (disposed)
                throw new ObjectDisposedException("Order", "Объект уже освобожден");
            return dateOrder;
        }

        public string GetPayWay()
        {
            if (disposed)
                throw new ObjectDisposedException("Order", "Объект уже освобожден");
            return payWay;
        }

        public string GetShopOrder()
        {
            if (disposed)
                throw new ObjectDisposedException("Order", "Объект уже освобожден");
            return shopOrder;
        }

        // Метод получения количества различных товаров в заказе
        public int GetItemCount()
        {
            if (disposed)
                throw new ObjectDisposedException("Order", "Объект уже освобожден");
            return itemCount;
        }

        // Метод получения товара по индексу с проверкой границ массива
        public Product GetOrderItem(int index)
        {
            if (disposed)
                throw new ObjectDisposedException("Order", "Объект уже освобожден");

            if (index >= 0 && index < itemCount)
            {
                return orderItems[index];
            }
            return null;
        }

        // Метод получения количества товара по индексу с проверкой границ массива
        public int GetQuantity(int index)
        {
            if (disposed)
                throw new ObjectDisposedException("Order", "Объект уже освобожден");

            if (index >= 0 && index < itemCount)
            {
                return quantities[index];
            }
            return 0;
        }


        public void Close()
        {
            Dispose();
        }
    }
}
